### ISSUE TITLE
Inlining in shared generic code leaves unused parts of dictionary lookups around

### ISSUE THREAD

**Issue by VSadov** [dotnet/coreclr#1691]
``` C#

    class Program
    {
        static void Main(string[] args)
        {
            new Generic<long, object>().test();
        }
    }

    class Generic<T, U>
    {
        private struct S1
        {
            public T field;
            public object stuff;
        }

        private S1[] arr = new S1[43];

        private static int GetLength(S1[] arr)
        {
            // last element is special and does not count towards length
            return arr.Length - 1;
        }

        public virtual void test()
        {
            System.Console.WriteLine(GetLength(arr));  // <-- stop here
        }

    }
```

The body of test() is:

```
            System.Console.WriteLine(GetLength(arr));  // <-- stop here
00000000  push        rsi 
00000001  sub         rsp,30h                                        
00000005  mov         qword ptr [rsp+28h],rcx        // ???????
0000000a  mov         rsi,rcx 
0000000d  mov         rcx,qword ptr [rsi+8] 
00000011  mov         ecx,dword ptr [rcx+8] 
00000014  dec         ecx 
00000016  call        0000000052CBEBF0 
0000001b  nop 
        }
0000001c  add         rsp,30h 
00000020  pop         rsi 
00000021  ret 
```

If I make GetLength an instance method, the strange store disappears and stack frame is smaller.

```
            System.Console.WriteLine(GetLength(arr));  // <-- stop here
00000000  sub         rsp,28h 
00000004  mov         rcx,qword ptr [rcx+8] 
00000008  mov         ecx,dword ptr [rcx+8] 
0000000b  dec         ecx 
0000000d  call        0000000052CBEC00 
00000012  nop 
        }
00000013  add         rsp,28h 
00000017  ret 
```

It seems unnatural that I can improve CQ by making static methods into instance methods in scenario like this, but it is very tempting when the pattern is on a hot path.

**Comment by mikedn** [comment]
It looks like lvaGenericsContextUsed is set when the call to GetLength is imported because GetLength is static and it needs the generic context argument (even if it doesn't use it). After inlining the compiler figures out that the whole generic context thing isn't needed and discards the code but it doesn't reset lvaGenericsContextUsed. In the end genFnProlog uses lvaGenericsContextUsed and forces this to be spilled.

**Comment by AndyAyersMS** [comment]
I have a prototype that gets this down to
```asm
G_M16884_IG01:
       4883EC28             sub      rsp, 40

G_M16884_IG02:
       488B4108             mov      rax, gword ptr [rcx+8]
       8B4008               mov      eax, dword ptr [rax+8]
       FFC8                 dec      eax

G_M16884_IG03:
       4883C428             add      rsp, 40
       C3                   ret
```
There's a phase ordering issue that remains: morph sees a helper call and records that the jit it needs to set up a minimal frame. The call later gets optimized away but the frame size does not get recalculated post optimization.

**Comment by AndyAyersMS** [comment]
Update with the call to `WriteLine` restored:
```asm
G_M16884_IG01:
       4883EC28             sub      rsp, 40

G_M16884_IG02:
       488B4908             mov      rcx, gword ptr [rcx+8]
       8B4908               mov      ecx, dword ptr [rcx+8]
       FFC9                 dec      ecx
       48B8D80167CAF97F0000 mov      rax, 0x7FF9CA6701D8

G_M16884_IG03:
       4883C428             add      rsp, 40
       48FFE0               rex.jmp  rax
```
Still don't need that frame.

**Comment by AndyAyersMS** [comment]
If we redo the outgoing arg size after optimization to reflect the calls that remain (say in `fgSimpleLowering` where we already walk all the trees) we end up with this:
```asm
G_M16884_IG01:

G_M16884_IG02:
       488B4908             mov      rcx, gword ptr [rcx+8]
       8B4908               mov      ecx, dword ptr [rcx+8]
       FFC9                 dec      ecx
       48B8D801A2C5F97F0000 mov      rax, 0x7FF9C5A201D8

G_M16884_IG03:
       48FFE0               rex.jmp  rax
```
Not sure yet if this is robust to do generally. Also won't work as easily for Unix ABI because we do the complex struct passing stuff in morph and lose some context.

**Comment by benaadams** [comment]
Also only 2 of the 3 `mov`s are required? Or is that a different issue?

**Comment by AndyAyersMS** [comment]
This should be fixed for the most part by #9756. Examples to the contrary welcomed.
